backend/CircleForms/Controllers/PostsController.cs

```diff
@@ -70,9 +76,112 @@ public async Task<PostRedis[]> GetCached()
     }
 
     [HttpGet("/posts/{id}")]
-    public async Task<PostRedis> GetCachedPost(string id)
+    public async Task<Post> GetPostForUser(string id)
     {
-        return await _postRepository.GetCached(id);
+        var post = await _postRepository.Get(id);
+        foreach (var question in post.Questions)
+        {
+            question.Answers = null;
+        }
+
+        return post;
+    }
+
+    private static readonly ConcurrentDictionary<string, SemaphoreSlim> _answerLockDictionary = new();
+    private static readonly Func<string, SemaphoreSlim> _semaphoreFactory = _ => new SemaphoreSlim(1);
+
+    [Authorize]
+    [HttpPost("/posts/{id}/answer")]
+    public async Task<IActionResult> Answer(string id, [FromBody] List<AnswerContract> answers)
+    {
+        var claim = HttpContext.User.Identity?.Name;
+        if (string.IsNullOrEmpty(claim) || !long.TryParse(claim, out var userId))
+        {
+            _logger.LogWarning("User had an invalid name claim on answer: {Claim}", claim);
+
+            return BadRequest();
+        }
+
+        var answerLock = _answerLockDictionary.GetOrAdd(id, _semaphoreFactory);
+        await answerLock.WaitAsync();
+
+        try
```

Comment:
I feel like this doesn't belong to the controller - maybe add [Automapper](https://automapper.org/) and do proper mapping for it instead, or at least move out into a separate service?